src/libexpr/primops.cc

```diff
@@ -2750,6 +2750,109 @@ static RegisterPrimOp primop_catAttrs({
     .fun = prim_catAttrs,
 });
 
+static void prim_functionStrict(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    state.forceValue(*args[0], pos);
+    if (args[0]->isPrimOpApp() || args[0]->isPrimOp()) {
+        v.mkNull();
+        return;
+    }
+
+    state.forceFunction(*args[0], pos, "while evaluating the argument passed to builtins.functionStrict");
+    const auto fun = args[0]->lambda.fun;
+
+    if (args[0]->isLambda() && fun->hasFormals()) {
+        v.mkBool(true);
+    } else {
+        v.mkNull();
+    }
+}
+
+static RegisterPrimOp primop_functionStrict({
+    .name = "__functionStrict",
+    .args = {"f"},
+    .doc = R"(
+      Return `true` if *f* is a function that is defined using strict function
+      syntax, which is to say, using an argument attribute set declaration such
+      as `{ a }:` or `{ ... }:`.
+
+      Proper strictness analysis is undecidable, so for all other functions the
+      return value is `null`. `false` is not returned.
+    )",
+    .fun = prim_functionStrict,
+});
+
+static void prim_functionOpen(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    state.forceValue(*args[0], pos);
+    if (args[0]->isPrimOpApp() || args[0]->isPrimOp()) {
+        v.mkNull();
+        return;
+    }
+
+    state.forceFunction(*args[0], pos, "while evaluating the argument passed to builtins.functionOpen");
+    const auto fun = args[0]->lambda.fun;
+
+    if (args[0]->isLambda()) {
+        v.mkNull();
+    }
+
+    if (fun->hasFormals()) {
+        v.mkBool(fun->formals->ellipsis);
+    } else {
+        v.mkNull();
+    }
+}
+
+static RegisterPrimOp primop_functionOpen({
+    .name = "__functionOpen",
+    .args = {"f"},
+    .doc = R"(
+      Return `true` if *f* is a function that is defined using the ellipsis syntax, such as `{ ... }:` or `{ foo, ... }:`.
+
+      Return `false` for functions defined with an attribute list but no ellipsis, such as `{ foo, bar }:`.
+
+      Return `null` for functions defined using plain lambdas, such as `x: ...`, as well as for built-in functions.
+    )",
+    .fun = prim_functionOpen,
+});
+
+static void prim_functionBindsAllAttrs(EvalState & state, const PosIdx pos, Value * * args, Value & v)
```

Comment:
What is this needed for? I don't really see an use case for this right away.

Reply:
It's used in this PR to report a warning when a flake's `outputs` function is invoked. This PR adds an attribute, which causes somewhat of a problem when the function is of the form

    args@{ self }:

The lack of ellipsis means that we'd have to remove the attribute for compatibility, while `args@` exposes that workaround, which would be a source of confusion. Detecting this situation and reporting a warning means that we get a reasonable upgrade story for this feature.

Reply:
Why not just check for an ellipsis (functionOpen)? Whether the user also binds the whole attribute set, is not really interesting for that.

Reply:
If it has ellipsis, but no binding for the whole attrset, there's no conflict and the backcompat logic can comply with the old interface and it doesn't have to warn about anything.

I don't want it to warn about something that isn't a problem, because too many warnings lead people to ignore them.

Reply:
Right, but why not just warn if it lacks an ellipsis. The full attribute set binding should be irrelevant.

Reply:
> The full attribute set binding should be irrelevant.

The ambiguity of https://github.com/NixOS/nix/pull/8908#discussion_r1315218936 does not occur if the full attrset binding is missing. So to keep warnings to a minimum, I want to know whether the full attrset binding exists or not.

I can agree that it *should* but unfortunately it would require me to warn in cases where it doesn't matter.